### PR TITLE
Switch to ubuntu base and fix bundler config -- Fixes "missing gem" errors

### PR DESCRIPTION
**Pull request by thenonsequitur**
### Purpose

The dockerhub ruby base image that our base image derives from does some opinioned bundler config that is not under our control. During a recent update to that dockerhub base image, they changed the bundler configs in a way that conflicts with our own desired setup, and has caused a lot of problems lately.

Most notably, since `BUNDLE_BIN` was removed in that last update, binstubs were not being properly generated, which has caused a slew of "gem not found" errors. But adding `BUNDLE_BIN` back into our image didn't solve the binstubs issue because of some other bundler config that was changed in the same update. The end result is that we no longer had a good way to generate binstubs under this new base image.

So this PR removes our dependence on the dockerhub ruby base image. Its bundler config is too opinionated and conflicts with our own, and as we've recently learned the hard way, it's liable to introduce breaking changes with no warning.

Instead, we now derive from `ubuntu:16.04`, install ruby manually, and set up the bundler config exactly how we want. The ruby installation steps themselves were largely based on the dockerhub ruby [Dockerfile](https://github.com/docker-library/ruby/blob/master/2.3/stretch/Dockerfile) (slightly modified for our use case and adapted to a bash script).

### Tests and risks

The ruby base image we used to derive from was in turn derived from the Debian stretch image, so this new image is using a different distro. Ubuntu 16.04 is based on Debian stretch, and 99% of things should be identical, but this does have the potential to cause breakage. As such, our base image version was bumped a major version to `3.0.0`.

This was tested locally. It was specifically tested against volume, hymnal, and vox-autotune (all locally modified to remove the stopgap hacks we've put in) to verify that it indeed solves the recent "missing gem" errors we've been seeing.

### Launch plan

1. Merge and CD as normal. 
2. In a hymnal branch, update to base image 3.0.0, verify that rubocop passes (it's currently failing every build). If so, merge/deploy that.
3. Next test against volume and vox-autotune. In a branch for each, remove the stopgap hacks we've put in and upgrade the base image. Verify that it fixes.
4. If that all goes well, update the base image for sbn and all systems-team-owned apps.
5. Put out an engineering announcement directing all repo owners to upgrade to base image 3.0.0.

### CODE REVIEW

File: ruby/build_scripts/install_ruby.sh [hunk 1]
```diff
@@ -0,0 +1,58 @@
+#!/bin/bash
+set -e
+
+MAJOR_VERSION=${1:-2.5}
+
+# Get the latest minor version of the given MAJOR_VERSION by parsing ruby-lang.org releases page
+version_info=$(curl -s https://www.ruby-lang.org/en/downloads/releases/ | grep "Ruby ${MAJOR_VERSION}" | head -1)
+if [[ ${version_info} =~ ${MAJOR_VERSION}\.[[:digit:]]+ ]]; then
+  RUBY_VERSION=${BASH_REMATCH}
```
Comment (thenonsequitur) [inline ruby/build_scripts/install_ruby.sh]: I'd specifically like some feed back on this. Is it appropriate for this script to be choosing the most recent patch level for a given major version? Or should we be locking to exact versions?

File: ruby/build_scripts/install_ruby.sh [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+else
```
Comment (thenonsequitur) [inline ruby/build_scripts/install_ruby.sh]: I'd specifically like some feed back on this. Is it appropriate for this script to be choosing the most recent patch level for a given major version? Or should we be locking to exact versions?

The argument for exact versions is that any patch upgrade will require rebuilding the base image anyway, so we could lock it and properly version upgrades of it. The argument for leaving it automatic is that as future unrelated upgrades are made and new base images released, we automatically get ruby patch upgrades in the process and don't ever have to think about manually updating them.

Not sure which way is better.

Comment (cschiewek) [reply ruby/build_scripts/install_ruby.sh]: I prefer the later, but I'm not married to the idea, I see the benefits of being more explicit too.

File: ruby/Dockerfile [hunk 3]
```diff
@@ -25,3 +28,15 @@ RUN chmod -R +x /opt/build/*
...
+ENV PATH=${GEM_HOME}/bin:${PATH} \
+  BUNDLE_PATH=${GEM_HOME} \
+  BUNDLE_BIN=${GEM_HOME}/bin/ \
+  BUNDLE_APP_CONFIG=/root/.bundle \
+  BUNDLE_SILENCE_ROOT_WARNING=1 \
+  BUNDLE_JOBS=4
+RUN mkdir -p ${GEM_HOME} \
+  && chmod 777 ${GEM_HOME} \
+  && /opt/build/install_ruby.sh ${RUBY_VERSION} \
+  && gem install bundler -v 1.16.2
```
Comment (cschiewek) [inline ruby/Dockerfile]: Should we maybe move this bundler configuration _inside_ the `install_ruby` script?

Comment (cschiewek) [reply ruby/Dockerfile]: Given that it's mostly `ENV` calls that's probably not possible, ignore me ;)

Comment (thenonsequitur) [reply ruby/Dockerfile]: yeah I originally had it in the install_ruby script, but realized there is not good way to transfer that env more generally unless it is in the Dockerfile, and as you pointed out, most of the bundler setup is the env vars.

File: ruby/build_scripts/install_ruby.sh (hunk 2 above)
Comment (keaty) [reply ruby/build_scripts/install_ruby.sh]: I prefer the latter as well, but I could be convinced otherwise.

Comment (keaty) [reply ruby/build_scripts/install_ruby.sh]: (As I recall, Ruby is pretty good about patch versions including only updates appropriate for patches)

**Comment by thomsbg** [comment]
(For the curious) why the change to base on ubuntu instead of debian? And if we're willing to take that risk, why not go all the way and base on alpine instead?

**Comment by thenonsequitur** [comment]
Debian and Ubuntu are practically identical in all ways that matter to us. Switching to alpine would be a much bigger endeavor. As for using Ubuntu instead of Debian, I think that's more just about familiarity among devs (it also simplifies the mysql client package management, which is more sane on ubuntu vs debian).

**Pull request merged**